### PR TITLE
Feature/present benchmarking

### PR DESCRIPTION
**Pull request by qh681248**
### PR Type

- Documentation content changes

### Description

Present the results of benchmarking in documentation.

### How Has This Been Tested?

Existing tests pass as expected.

No new tests were introduced
### Checklist before requesting a review
- [x] I have made sure that my PR is not a duplicate.
- [x] My code follows the style guidelines of this project.
- [x] I have ensured my code is easy to understand, including docstrings and comments where necessary.
- [x] I have performed a self-review of my code.
- [x] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] New and existing unit tests pass locally with my changes.
- [x] Any dependent changes have been merged and published in downstream modules.
- [x] I have updated CHANGELOG.md, if appropriate.

### CODE REVIEW

File: documentation/source/benchmark.rst [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+
+1. **Dataset**: A synthetic dataset of 1_000 points is generated to test the
+   quality of coreset algorithms.
+
+2. **Coreset Generation**: Coresets of different sizes (10, 50, 100, and 200 points)
+   are generated using each coreset algorithm.
+
+3. **Evaluation Metrics**: Two metrics evaluate the quality of the generated coresets:
+   - **Maximum Mean Discrepancy (MMD)**
+   - **Kernel Stein Discrepancy (KSD)**
```
Comment (rg936672) [inline documentation/source/benchmark.rst:74]: Can these empirical metrics be calculated for the coresets from the other three tests? (Happy for this to be out of scope if it's a lot of extra work!)

File: documentation/source/benchmark.rst [hunk 2]
```diff
@@ -0,0 +1,319 @@
...
+  algorithms on a synthetic dataset.
+
+Test 3: Benchmarking Coreset Algorithms on Pixel Data from an Image
+-------------------------------------------------------------------
+
+This test evaluates the performance of coreset algorithms on pixel data extracted
+from an input image. The process follows these steps:
+
+1. **Input Image**: An image is loaded and downsampled to reduce resolution (a
+   downsampling factor of 1 corresponds to no downsampling).
```
Comment (rg936672) [inline documentation/source/benchmark.rst:234]: Why mention the downsampling if it's not then used in the test?

File: documentation/source/benchmark.rst [hunk 3]
```diff
@@ -0,0 +1,319 @@
...
+
+  **Figure 2**: Time taken to generate coreset for each coreset algorithm. Bar heights
+  represent the average time taken. Error bars represent the min-max range for each
+  coreset size across 5 runs.
+
+Test 2: Benchmarking Coreset Algorithms on a Synthetic Dataset
+--------------------------------------------------------------
+
+In this second test, we evaluate the performance of the coreset algorithms on a
+**synthetic dataset**. The dataset consists of 1_000 two-dimensional points,
```
Comment (rg936672) [inline documentation/source/benchmark.rst:63]: Rephrase something more like as "points in two-dimensional space", or "two-dimensional vectors"? As to me a point is typically zero-dimensional!

File: examples/benchmarking_images/mnist_benchmark_time_taken.png
Comment (rg936672) [inline examples/benchmarking_images/mnist_benchmark_time_taken.png:1]: Would it make more sense to use a log scale here, to make it easier to compare the other three methods that aren't `SteinThinning`? (If you make changes to these images, I'd recommend moving them to Git LFS first, so that having multiple versions of the file in the history won't affect repo performance.)

File: documentation/source/benchmark.rst [hunk 4]
```diff
@@ -0,0 +1,319 @@
...
+   measured on the test set of 10_000 images.
+
+6. **Evaluation**: Due to randomness in the coreset algorithms and training process,
+   the experiment is repeated 5 times with different random seeds. The benchmark is run
+   on an **Amazon g4dn.12xlarge instance** with 4 NVIDIA T4 Tensor Core GPUs, 48 vCPUs,
+   and 192 GiB memory.
+
+**Results**:
+- Plots showing accuracy (with error bars) and time taken for each coreset size and
+algorithm.
```
Comment (rg936672) [inline documentation/source/benchmark.rst:43]: Should be a little clearer about what exactly the "accuracy" is here - I assume that this is something like accuracy of the MLP's predictions on the test set? If so, it'd be good to compare it against the accuracy of the same MLP trained on the full training set as a baseline.

**Review by rg936672 — CHANGES_REQUESTED** [review]
This looks good! A few queries and formatting changes. Before making any changes to the images, consider moving them to Git LFS first to avoid clogging up the repo.

File: benchmark/mnist_benchmark.py [hunk 5]
```diff
@@ -626,14 +628,14 @@ def main() -> None:
...
+        "weight_decay": 1e-4,
+        "patience": 15,
         "min_delta": 0.001,
     }
 
     # Run the experiment with 5 different random keys
     # pylint: disable=duplicate-code
-    for i in range(5):
+    for i in range(2):
         print(f"Run {i + 1} of 5:")
```
[the remark this answers is not in the document]
Comment (qh681248) [reply benchmark/mnist_benchmark.py:639]: Has been changed back to 5

File: benchmark/mnist_benchmark_coresets_only.py [hunk 6]
```diff
@@ -99,7 +99,6 @@ def main() -> None:
 
     # Run the experiment with 5 different random keys
     for i in range(5):
-        print(f"Run {i + 1} of 5:")
```
[the remark this answers is not in the document]
Comment (qh681248) [reply benchmark/mnist_benchmark_coresets_only.py:102]: It has been added back for consistency

File: documentation/source/benchmark.rst [hunk 7]
```diff
@@ -0,0 +1,319 @@
+Benchmarking Coreset Algorithms
+===============================
+
+In this benchmark, we assess the performance of four different coreset algorithms:
+:class:`coreax.solvers.KernelHerding`, :class:`coreax.solvers.SteinThinning`,
+:class:`coreax.solvers.RandomSample`, and :class:`coreax.solvers.RPCholesky`.
```
[the remark this answers is not in the document]
Comment (qh681248) [reply documentation/source/benchmark.rst:6]: Cheers, that makes a lot of sense.

File: documentation/source/benchmark.rst (hunk 4 above)
Comment (qh681248) [reply documentation/source/benchmark.rst:43]: Rephrased for clarity and also added the baseline accuracy for 97.31% Also added code in MNIST_benchmark.py to compute this accuracy.

File: documentation/source/benchmark.rst [hunk 8]
```diff
@@ -0,0 +1,319 @@
...
+
+In this second test, we evaluate the performance of the coreset algorithms on a
+**synthetic dataset**. The dataset consists of 1_000 two-dimensional points,
+generated using :func:`sklearn.datasets.make_blobs`. The process follows these steps:
+
+1. **Dataset**: A synthetic dataset of 1_000 points is generated to test the
+   quality of coreset algorithms.
+
+2. **Coreset Generation**: Coresets of different sizes (10, 50, 100, and 200 points)
+   are generated using each coreset algorithm.
```
[the remark this answers is not in the document]
Comment (qh681248) [reply documentation/source/benchmark.rst:70]: It was the former, now it has been run 5 times and averaged,

File: documentation/source/benchmark.rst (hunk 1 above)
Comment (qh681248) [reply documentation/source/benchmark.rst:74]: These metrics can be computed for the fourth test (video benchmarking) and it might be possible to compute them for the image test (on gpu machine for instance) but definitely not for the first test. I have left them out for now.

File: documentation/source/benchmark.rst (hunk 2 above)
Comment (qh681248) [reply documentation/source/benchmark.rst:234]: Removed any reference to downsampling

File: benchmark/mnist_benchmark_coresets_only.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (tp832944) [reply benchmark/mnist_benchmark_coresets_only.py]: @qh681248 Still outstanding.

File: tests/performance/compare.py [hunk 9]
```diff
@@ -94,6 +94,7 @@ def parse_args() -> Tuple[Path, Path, str, Path]:
     )
 
 
+# pylint: disable=R0801
```
[the remark this answers is not in the document]
Comment (tp832944) [reply tests/performance/compare.py]: Still outstanding.

File: tests/performance/compare.py [hunk 10]
```diff
@@ -133,6 +134,7 @@ def date_from_filename(path: Path) -> Optional[Tuple[datetime.datetime, str]]:
     ), git_hash
 
 
+# pylint: disable=R0801
```
[the remark this answers is not in the document]
Comment (tp832944) [reply tests/performance/compare.py]: Still outstanding.

File: examples/benchmarking_images/mnist_benchmark_time_taken.png
Comment (tp832944) [reply examples/benchmarking_images/mnist_benchmark_time_taken.png:1]: Still outstanding.

File: tests/performance/compare.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (tp832944) [reply tests/performance/compare.py]: @qh681248 Still outstanding. If we do need to keep it, please replace R0801 with `duplicate-code`.

File: tests/performance/compare.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (tp832944) [reply tests/performance/compare.py]: @qh681248 Still outstanding. See comment on line 97.
